akka-stream/src/main/scala/akka/stream/impl/SubscriberManagement.scala

```diff
@@ -126,8 +126,16 @@ private[akka] trait SubscriberManagement[T] extends ResizableMultiReaderRingBuff
                   // if we are at end-of-stream and have nothing more to read we complete now rather than after the next `requestMore`
                   if ((eos ne NotReached) && buffer.count(subscription) == 0) Long.MinValue else 0
                 } else if (buffer.count(subscription) > 0) {
-                  subscription.dispatch(buffer.read(subscription)) // FIXME this does not gracefully handle the case if onNext throws
-                  dispatchFromBufferAndReturnRemainingRequested(requested - 1, eos)
+                  val goOn = try {
+                    subscription.dispatch(buffer.read(subscription))
+                    true
+                  } catch {
+                    case _: SpecViolation ⇒
+                      unregisterSubscriptionInternal(subscription)
+                      false
+                  }
+                  if (goOn) dispatchFromBufferAndReturnRemainingRequested(requested - 1, eos)
+                  else Long.MinValue
```

Comment:
this is the contract of the method; tests pass, but we will remove this whole file (and a few others) soon I hope